clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4288,6 +5343,41 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
   }
 }
 
+void ConstructFreeFunctionKernel(Sema &SemaRef, FunctionDecl *FD) {
+  SyclKernelArgsSizeChecker argsSizeChecker(SemaRef, FD->getLocation(),
+                                            false /*IsSIMDKernel*/);
+  SyclKernelDeclCreator kernel_decl(SemaRef, FD->getLocation(), FD->isInlined(),
+                                    FD);
+
+  // TODO: remove this dummy struct
+  RecordDecl *DummyClass =
```

Comment:
I don't understand what is happening here. Why `DummyClass `and `DummyClass1`. Why not just nullptr if you don't need this?

Reply:
The problem here is VarEntity, a member of SyclKernelBodyCreator. The compiler insists that the constructor initialize VarEntity. For free functions we don't use VarEntity. If you can tell me how to provide an init for it, we won't need this dummy variable.

Reply:
You need to conditionalize the code to not create the VarEntity/InitListExprs/KernelObjClone/etc because it does not make sense to do so for free functions.